src/storage/index/index_factory.cpp

```diff
@@ -0,0 +1,77 @@
+#include "storage/index/index_factory.h"
+#include <utility>
+
+namespace terrier::storage::index {
+Index *IndexFactory::Build() const {
+  TERRIER_ASSERT(!key_schema_.empty(), "Cannot build an index without a KeySchema.");
+  TERRIER_ASSERT(constraint_type_ != ConstraintType::INVALID, "Cannot build an index without a ConstraintType.");
+
+  IndexMetadata metadata(key_schema_);
+
+  // figure out if we can use CompactIntsKey
+  bool use_compact_ints = true;
+  uint32_t key_size = 0;
+
+  for (uint16_t i = 0; use_compact_ints && i < key_schema_.size(); i++) {
+    const auto &attr = key_schema_[i];
+    use_compact_ints = use_compact_ints && !attr.IsNullable() && CompactIntsOk(attr.GetType());  // key type ok?
+    key_size += type::TypeUtil::GetTypeSize(attr.GetType());
+    use_compact_ints = use_compact_ints && key_size <= sizeof(uint64_t) * INTSKEY_MAX_SLOTS;  // key size fits?
```

Comment:
[Minor] Hope you don't mind that I feel like these three lines are a little bit cumbersome. But it's a really tiny problem :)

Reply:
Actually, it is Wen's code.

Reply:
"Wan"